README.md
# Dark SynthWave '84 - VS Code theme

<p align="center">
  <a href="https://marketplace.visualstudio.com/items?itemName=lukinco.lukin-vscode-theme">
    <img alt="Version" src="https://vsmarketplacebadge.apphb.com/version/aaronYoung.dark-synthwave-vscode.svg" />
  </a>
  <a href="https://marketplace.visualstudio.com/items?itemName=aaronYoung.dark-synthwave-vscode">
    <img alt="Downloads" src="https://vsmarketplacebadge.apphb.com/downloads/aaronYoung.dark-synthwave-vscode.svg" />
  </a>
  <a href="https://marketplace.visualstudio.com/items?itemName=aaronYoung.dark-synthwave-vscode">
    <img alt="Installs" src="https://vsmarketplacebadge.apphb.com/installs/aaronYoung.dark-synthwave-vscode.svg" />
  </a>
</p>

This is an amalgamation between two themes, [Lukin Theme](https://marketplace.visualstudio.com/items?itemName=lukinco.lukin-vscode-theme)
& [Synthwave '84](https://marketplace.visualstudio.com/items?itemName=RobbOwen.synthwave-vscode)
plus a few tweaks of my own.

Do you like the theme Synthwave '84? Do you wish there was a dark version of it?

I present Dark SynthWave '84, which has two variations, with and without the glow effect with the latter being the default.

![Neon glowing text](https://raw.githubusercontent.com/brainomite/dark-synthwave-vscode/main/theme.png)

<p class="figcaption" style="text-align: center; margin-top: -0.5rem; margin-bottom: 2rem; font-size: smaller">
  Glow enabled
</p>

![Neon text, no glow](https://raw.githubusercontent.com/brainomite/dark-synthwave-vscode/main/theme-no-glow.png)

<p class="figcaption" style="text-align: center; margin-top: -0.5rem; margin-bottom: 2rem; font-size: smaller">
  Glow disabled
</p>

## But...why?
I wanted something easier on my eyes than the original Synthwave '84.

## I love your font!
Actually its two fonts! they are [JetBrains Mono](https://www.jetbrains.com/lp/mono/")
& [Victor Mono](https://rubjo.github.io/victor-mono/) see my blog
post on [how I used two fonts simultaneously](https://aaronyoung.dev/blog/2021-01-24-vscode-dual-fonts/)


## Installation
To begin with, [install the base theme from the VS Marketplace](https://marketplace.visualstudio.com/items?itemName=AaronYoung.dark-synthwave-vscode). This is the way Dark SynthWave '84 is intended to be used day-to-day. If you want to enable the gratuitous 80s glow, it needs a little extra work to get it going.

### Disclaimer
VS code doesn't natively support text effects and as a result, the glow is experimental. It's likely to be buggy and, whilst it looks rad, it isn't intended for extended use. To enable the glow, the extension has to modify the internal files of VS Code, so use with caution. Should something go wrong, you can disable the glow by following the instructions below. If for any reason you can't open VS Code, you can fix the issue with a fresh install of VS Code.

If you do decide to use the glow effect, you do so at your own risk.

### To enable the glow

Firstly, if you are a Windows user, you may need to run VS Code with administrator privileges. For Linux and Mac users, Code must not be installed in a read-only location and you must have write permissions.

To activate the glow, Set your active colour theme to Dark SynthWave '84 - as now the glow is only active when the base theme is selected. Open your command palette with `Ctrl + Shift + P` or `Shift + ⌘ + P` and choose "__Dark Synthwave '84: Enable Neon Dreams__". It will prompt you to restart, and when you do the lights should be on 😲

#### To customize the glow brightness
Go to __Settings>extensions>Dark Synthwave84: Brightness__ and set the value.

The value should be a _float value_ from 0 to 1, where 0.0 is fully transparent. The default brightness is 0.45. To avoid eye strain, avoid using higher brightness values for extended periods of time.

To see the changes, you need to rerun the activation function. Open your command palette with `Ctrl + Shift + P` or `Shift + ⌘ + P` and choose "__Dark Synthwave '84: Enable Neon Dreams__".

Note: Changing the brightness currently only affects the opacity of the glow, the text will remain white (that may change in future updates). If you want to disable the glow effect but retain the chrome updates, see below.

#### To enable editor chrome updates, but disable glow
Go to __Settings>extensions>Dark Synthwave84: Disable Glow__ and check off the option to disable the glow effect.

To see the changes, you need to rerun the activation function. Open your command palette with `Ctrl + Shift + P` or `Shift + ⌘ + P` and choose "__Enable Neon Dreams__".

### To remove corruption warning and [unsupported] from title-bar
Because enabling the glow modifies core files, VS code will interpret this as the core being 'corrupted' and you may see an error message on restarting your editor. You can safely dismiss this message, or remove it entirely with the [Fix VSCode Checksums](https://marketplace.visualstudio.com/items?itemName=lehni.vscode-fix-checksums 'Fix VSCode Checksums') extension.

Upon installation of 'Fix VSCode Checksums', open the command palette and execute `Fix Checksums: Apply`. You will need to completely restart VSCode after execution, reopening without fully exiting might not be enough.

## Updates
Every time you update VS code, you will need to repeat this step to re-enable the glow.

## Disabling the glow and uninstalling
The glow effect started as a joke and was never intended for long-term coding sessions. If you want to turn it off, you can disable it at any time by opening your command palette with `Ctrl + Shift + P` or `Shift + ⌘ + P` and choose "__Disable Neon Dreams__".

## Compatibility
This theme is still **very much a work in progress**. I primarily develop in HTML & CSS, JS, React so, whilst those language sets should look pretty good, there will likely be issues for other languages. I'll work on adding more support as I go. If you find anything glaringly wrong, raise an issue and I'll try to fix it as soon as I can.

## Known Issues

- Some computers, with the glow enabled, have sluggishness when scrolling. This depends on the capability of the machine, newer machines can render faster than older ones.

## Contributing
I'm really happy to consider any contributions to this theme. Before you make any changes, [please read the contribution guide](https://github.com/brainomite/dark-synthwave-vscode/blob/master/CONTRIBUTING.md).

## Thanks
Lastly, I couldn't have made this if it weren't for the fantastic work of[Robb Owen](https://marketplace.visualstudio.com/publishers/RobbOwen) for their theme [Synthwave '84](https://marketplace.visualstudio.com/items?itemName=RobbOwen.synthwave-vscode) theme & [Lukin Co](https://marketplace.visualstudio.com/publishers/lukinco) for their [Lukin Theme](https://marketplace.visualstudio.com/items?itemName=lukinco.lukin-vscode-theme). 🙏



CHANGELOG.md
# Change Log
All notable changes to the "Dark Synthwave '84"
extension pack will be documented in this file.

The format is based on [Keep ]a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Seman]tic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.14] - 2021-01-26
### Add
- Icon for project
## [0.0.13] - 2021-01-26
### Update
- Add badges to README
- Add dates to CHANGELOG
## [0.0.12] - 2021-01-26
### Add
- Scripts to package.json
## [0.0.11] - 2021-01-26
__NO CHANGES__
This was an error.
## [0.0.10] - 2021-01-26
### Update
- Fix grammar issues in README
- Add known issues section
## [0.0.9] - 2021-01-26
### Update
- Indicate that non-glow is the default
## [0.0.8] - 2021-01-26
### Remove
- Remove non-functional image compare
### Add
- To README, image without glow & instructions on font
## [0.0.7] - 2021-01-26
### Update
- Add image comparison

## [0.0.6] - 2021-01-26
### Update
- Update theme image
### Added
- theme image without glow
## [0.0.5] - 2021-01-26
### Update
- Fix error for instructions in the README
## [0.0.4] - 2021-01-26
### Update
- Change theme.png
- Update license
- Add license to package

## [0.0.3] - 2021-01-26
### Update
- Change README & theme image

## [0.0.2] - 2021-01-26
### Added
- This change log
### Updated
- Fix broken image link in README

## [0.0.1] - 2021-01-26
Initial public preview




src/js/theme_template.js
(function () {

  // Grab body node
  const bodyNode = document.querySelector('body');

  // Replace the styles with the glow theme
  const initNeonDreams = (disableGlow, obs) => {
    var themeStyleTag = document.querySelector('.vscode-tokens-styles');

    if (!themeStyleTag) {
      return;
    }

    var initialThemeStyles = themeStyleTag.innerText;

    var updatedThemeStyles = initialThemeStyles;

    if (!disableGlow) {
      /* replace neon red */
      updatedThemeStyles = updatedThemeStyles.replace(/color: #fe4450;/g, "color: #fff5f6; text-shadow: 0 0 2px #000, 0 0 10px #fc1f2c[NEON_BRIGHTNESS], 0 0 5px #fc1f2c[NEON_BRIGHTNESS], 0 0 25px #fc1f2c[NEON_BRIGHTNESS];");

      /* replace neon pink */
      updatedThemeStyles = updatedThemeStyles.replace(/color: #ff7edb;/g, "color: #f92aad; text-shadow: 0 0 2px #100c0f, 0 0 5px #dc078e33, 0 0 10px #fff3;");

      /* replace yellow */
      updatedThemeStyles = updatedThemeStyles.replace(/color: #fede5d;/g, "color: #f4eee4; text-shadow: 0 0 2px #393a33, 0 0 8px #f39f05[NEON_BRIGHTNESS], 0 0 2px #f39f05[NEON_BRIGHTNESS];");

      /* replace green */
      updatedThemeStyles = updatedThemeStyles.replace(/color: #72f1b8;/g, "color: #72f1b8; text-shadow: 0 0 2px #100c0f, 0 0 10px #257c55[NEON_BRIGHTNESS], 0 0 35px #212724[NEON_BRIGHTNESS];");

      /* replace blue */
      updatedThemeStyles = updatedThemeStyles.replace(/color: #36f9f6;/g, "color: #fdfdfd; text-shadow: 0 0 2px #001716, 0 0 3px #03edf9[NEON_BRIGHTNESS], 0 0 5px #03edf9[NEON_BRIGHTNESS], 0 0 8px #03edf9[NEON_BRIGHTNESS];");
    }

    /* append the remaining styles */
    updatedThemeStyles = `${updatedThemeStyles}[CHROME_STYLES]`;

    const newStyleTag = document.createElement('style');
    newStyleTag.setAttribute("id", "synthwave-84-theme-styles");
    newStyleTag.innerText = updatedThemeStyles.replace(/(\r\n|\n|\r)/gm, '');
    document.body.appendChild(newStyleTag);

    console.log("Dark Synthwave '84: NEON DREAMS initialized!");

    // disconnect the observer because we don't need it anymore
    if (obs) {
      obs.disconnect();
    }
  };

  // Callback function to execute when mutations are observed
  const watchForBootstrap = function(mutationsList, observer) {
      for(let mutation of mutationsList) {
          if (mutation.type === 'attributes') {
            // only init if we're using a Synthwave 84 subtheme
            const isUsingSynthwave = document.querySelector(
              '[class*="AaronYoung-dark-synthwave-vscode-themes"]'
            );
            // does the style div exist yet?
            const tokensLoaded = document.querySelector('.vscode-tokens-styles');
            // does it have content ?
            const tokenStyles = document.querySelector('.vscode-tokens-styles').innerText;

            // sometimes VS code takes a while to init the styles content, so stop this observer and add an observer for that
            if (isUsingSynthwave && tokensLoaded) {
              observer.disconnect();
              observer.observe(tokensLoaded, { childList: true });
            }
          }
          if (mutation.type === 'childList') {
            const isUsingSynthwave = document.querySelector(
              '[class*="AaronYoung-dark-synthwave-vscode-themes"]'
            );
            const tokensLoaded = document.querySelector('.vscode-tokens-styles');
            const tokenStyles = document.querySelector('.vscode-tokens-styles').innerText;

            // Everything we need is ready, so initialise
            if (isUsingSynthwave && tokensLoaded && tokenStyles) {
                initNeonDreams([DISABLE_GLOW], observer);
            }
          }
      }
  };

  // try to initialise the theme
  initNeonDreams([DISABLE_GLOW]);

  // Use a mutation observer to check when we can bootstrap the theme
  const observer = new MutationObserver(watchForBootstrap);
  observer.observe(bodyNode, { attributes: true });

})();


src/extension.js
const path = require('path');
const fs = require('fs');
const vscode = require('vscode');
const diff = require('semver/functions/diff');

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
	this.extensionName = "AaronYoung.dark-synthwave-vscode";
	this.cntx = context;
	this.extension = vscode.extensions.getExtension(this.extensionName);
	if (this.extension) {
		// grab current version number
		this.version = this.extension.packageJSON.version;

		// grab last recorded version
		const prevVersion = context.globalState.get(`${this.extensionName}.version`);

		if (prevVersion) {
			// check it has changed.
			const d = diff(this.version, prevVersion);
			// show again on major or minor updates
			if (d == 'major' || d == 'minor') {
				showUpdatePage();
				context.globalState.update(`${this.extensionName}.version`, this.version);
			}
		} else {
			showUpdatePage();
			context.globalState.update(`${this.extensionName}.version`, this.version);
		}

	}

	const config = vscode.workspace.getConfiguration("darkSynthwave84");

	let disableGlow = config && config.disableGlow ? !!config.disableGlow : false;

	let brightness = parseFloat(config.brightness) > 1 ? 1 : parseFloat(config.brightness);
	brightness = brightness < 0 ? 0 : brightness;
	brightness = isNaN(brightness) ? 0.45 : brightness;

	const parsedBrightness = Math.floor(brightness * 255).toString(16).toUpperCase();
	let neonBrightness = parsedBrightness;

	let disposable = vscode.commands.registerCommand(
    "darkSynthwave84.enableNeon",
    function () {
      const isWin = /^win/.test(process.platform);
      const appDir = path.dirname(require.main.filename);
      const base = appDir + (isWin ? "\\vs\\code" : "/vs/code");

      const htmlFile =
        base +
        (isWin
          ? "\\electron-browser\\workbench\\workbench.html"
          : "/electron-browser/workbench/workbench.html");

      const templateFile =
        base +
        (isWin
          ? "\\electron-browser\\workbench\\neondreams.js"
          : "/electron-browser/workbench/neondreams.js");
      try {
        // const version = context.globalState.get(`${context.extensionName}.version`);

        // generate production theme JS
        const chromeStyles = fs.readFileSync(
          __dirname + "/css/editor_chrome.css",
          "utf-8"
        );
        const jsTemplate = fs.readFileSync(
          __dirname + "/js/theme_template.js",
          "utf-8"
        );
        const themeWithGlow = jsTemplate.replace(
          /\[DISABLE_GLOW\]/g,
          disableGlow
        );
        const themeWithChrome = themeWithGlow.replace(
          /\[CHROME_STYLES\]/g,
          chromeStyles
        );
        const finalTheme = themeWithChrome.replace(
          /\[NEON_BRIGHTNESS\]/g,
          neonBrightness
        );
        fs.writeFileSync(templateFile, finalTheme, "utf-8");

        // modify workbench html
        const html = fs.readFileSync(htmlFile, "utf-8");

        // check if the tag is already there
        const isEnabled = html.includes("neondreams.js");

        if (!isEnabled) {
          // delete synthwave script tag if there
          let output = html.replace(
            /^.*(<!-- SYNTHWAVE 84 --><script src="neondreams.js"><\/script><!-- NEON DREAMS -->).*\n?/gm,
            ""
          );
          // add script tag
          output = html.replace(
            /\<\/html\>/g,
            `	<!-- SYNTHWAVE 84 --><script src="neondreams.js"></script><!-- NEON DREAMS -->\n`
          );
          output += "</html>";

          fs.writeFileSync(htmlFile, output, "utf-8");

          vscode.window
            .showInformationMessage(
              "Neon Dreams enabled. VS code must reload for this change to take effect. Code may display a warning that it is corrupted, this is normal. You can dismiss this message by choosing 'Don't show this again' on the notification.",
              { title: "Restart editor to complete" }
            )
            .then(function (msg) {
              vscode.commands.executeCommand("workbench.action.reloadWindow");
            });
        } else {
          vscode.window
            .showInformationMessage(
              "Neon dreams is already enabled. Reload to refresh JS settings.",
              { title: "Restart editor to refresh settings" }
            )
            .then(function (msg) {
              vscode.commands.executeCommand("workbench.action.reloadWindow");
            });
        }
      } catch (e) {
        if (/ENOENT|EACCES|EPERM/.test(e.code)) {
          vscode.window.showInformationMessage(
            "You must run VS code with admin priviliges in order to enable Neon Dreams."
          );
          return;
        } else {
          vscode.window.showErrorMessage(
            "Something went wrong when starting neon dreams"
          );
          return;
        }
      }
    }
  );

	let disable = vscode.commands.registerCommand(
    "darkSynthwave84.disableNeon",
    uninstall
  );
  let whatsNew = vscode.commands.registerCommand(
    "darkSynthwave84.whatsNew",
    showUpdatePage
  );

	context.subscriptions.push(disposable);
	context.subscriptions.push(disable);
	context.subscriptions.push(whatsNew);
}
exports.activate = activate;


function showUpdatePage() {
		const panel = vscode.window.createWebviewPanel(
			`synthwave.whatsNew`, // Identifies the type of the webview. Used internally
			'What\'s new for Synthwave \'84', // Title of the panel displayed to the user
			vscode.ViewColumn.One, // Editor column to show the new webview panel in.
			{ enableScripts: !0 } // Webview options. More on these later.
		);

		const viewPath = path.join(this.cntx.extensionPath, "whats-new", "view.html");
		const viewResourcePath = panel.webview.asWebviewUri(viewPath);
		const htmlContent = fs.readFileSync(viewPath, "utf-8");
		panel.webview.html = htmlContent;
}

// this method is called when your extension is deactivated
function deactivate() {
	// ...
}

function uninstall() {
	var isWin = /^win/.test(process.platform);
	var appDir = path.dirname(require.main.filename);
	var base = appDir + (isWin ? "\\vs\\code" : "/vs/code");
	var htmlFile =
		base +
		(isWin
			? "\\electron-browser\\workbench\\workbench.html"
			: "/electron-browser/workbench/workbench.html");

	// modify workbench html
	const html = fs.readFileSync(htmlFile, "utf-8");

	// check if the tag is already there
	const isEnabled = html.includes("neondreams.js");

	if (isEnabled) {
		// delete synthwave script tag if there
		let output = html.replace(/^.*(<!-- SYNTHWAVE 84 --><script src="neondreams.js"><\/script><!-- NEON DREAMS -->).*\n?/mg, '');
		fs.writeFileSync(htmlFile, output, "utf-8");

		vscode.window
			.showInformationMessage("Neon Dreams disabled. VS code must reload for this change to take effect", { title: "Restart editor to complete" })
			.then(function(msg) {
				vscode.commands.executeCommand("workbench.action.reloadWindow");
			});
	} else {
		vscode.window.showInformationMessage('Neon dreams isn\'t running.');
	}
}

module.exports = {
	activate,
	deactivate
}
